solr/modules/llm/src/java/org/apache/solr/llm/embedding/SolrEmbeddingModel.java

```diff
@@ -0,0 +1,157 @@
+
+public class SolrEmbeddingModel implements Accountable {
+  private static final long BASE_RAM_BYTES =
+      RamUsageEstimator.shallowSizeOfInstance(SolrEmbeddingModel.class);
+  public static final String TIMEOUT_PARAM = "timeout";
+  public static final String LOG_REQUESTS_PARAM = "logRequests";
+  public static final String LOG_RESPONSES_PARAM = "logResponses";
+  public static final String MAX_SEGMENTS_PER_BATCH_PARAM = "maxSegmentsPerBatch";
+  public static final String MAX_RETRIES_PARAM = "maxRetries";
+
+  protected final String name;
+  private final Map<String, Object> params;
+  private EmbeddingModel embedder;
+  private Integer hashCode;
+
+  public static SolrEmbeddingModel getInstance(
+      String className, String name, Map<String, Object> params) throws EmbeddingModelException {
+    try {
+      EmbeddingModel embedder;
+      Class<?> modelClass = Class.forName(className);
+      var builder = modelClass.getMethod("builder").invoke(null);
+      if (params != null) {
+        for (String paramName : params.keySet()) {
+          switch (paramName) {
+            case TIMEOUT_PARAM:
+              Duration timeOut = Duration.ofSeconds((Long) params.get(paramName));
+              builder.getClass().getMethod(paramName, Duration.class).invoke(builder, timeOut);
+              break;
+            case LOG_REQUESTS_PARAM:
+              builder
+                  .getClass()
+                  .getMethod(paramName, Boolean.class)
+                  .invoke(builder, params.get(paramName));
+              break;
+            case LOG_RESPONSES_PARAM:
+              builder
+                  .getClass()
+                  .getMethod(paramName, Boolean.class)
+                  .invoke(builder, params.get(paramName));
+              break;
+            case MAX_SEGMENTS_PER_BATCH_PARAM:
+              builder
+                  .getClass()
+                  .getMethod(paramName, Integer.class)
+                  .invoke(builder, ((Long) params.get(paramName)).intValue());
+              break;
+            case MAX_RETRIES_PARAM:
+              builder
+                  .getClass()
+                  .getMethod(paramName, Integer.class)
+                  .invoke(builder, ((Long) params.get(paramName)).intValue());
+              break;
+            default:
+              builder
+                  .getClass()
+                  .getMethod(paramName, String.class)
+                  .invoke(builder, params.get(paramName));
+          }
+        }
+      }
+      embedder = (EmbeddingModel) builder.getClass().getMethod("build").invoke(builder);
+      return new SolrEmbeddingModel(name, embedder, params);
+    } catch (final Exception e) {
+      throw new EmbeddingModelException("Model loading failed for " + className, e);
+    }
+  }
+
+  public SolrEmbeddingModel(String name, EmbeddingModel embedder, Map<String, Object> params) {
+    this.name = name;
+    this.embedder = embedder;
+    this.params = params;
```

Comment:
minor: it seems `embedder` could be `final` too and then the hash code could be calculated in the constructor?
```suggestion
    this.params = params;
    this.hashCode = calculateHashCode();
```